### ISSUE TITLE
eqeqeq rule allow-undefined

### ISSUE THREAD

**Issue by SyMind** [eslint/eslint#13537]
**What rule do you want to change?**

Require === and !== (eqeqeq)
https://eslint.org/docs/rules/eqeqeq#allow-null

**Does this change cause the rule to produce more or fewer warnings?**

In ECMA262:

> null value is primitive value that represents the intentional absence of any object value.

> undefined value is primitive value used when a variable has not been assigned a value.

In TypeScript team coding guidelines:

> Use undefined. Do not use null.

https://github.com/microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined

So i think comparing with the undefined literal should also can be allowed.

**How will the change be implemented? (New option, new default behavior, etc.)?**
Please  support allow-undefined.

**Please provide some example code that this change will affect:**

```js
foo == null
foo == undefined
```

**What does the rule currently do for this code?**

```js
foo == null // correct
foo == undefined // incorrect
```

**What will the rule do after it's changed?**

```js
foo == null // correct
foo == undefined // correct
```

**Are you willing to submit a pull request to implement this change?**

It's my pleasure

**Comment by g-plane** [comment]
Please note that the guideline is for TypeScript contributors, not for users.

**Comment by mdjermanovic** [comment]
Makes sense to me, but I guess this would be `["error", "always", { "undefined": "ignore" }]`, rather than `allow-undefined`.

`allow-null` is deprecated in favor of `"null": "always" | "never" | "ignore"`.

This is documented at the end of the [always](https://eslint.org/docs/rules/eqeqeq#always-9) section, and in the [allow-null](https://eslint.org/docs/rules/eqeqeq#allow-null) section. It feels that the document is missing examples or a whole new section for the `"null"` option.

The new option for `undefined`, [if accepted](https://eslint.org/docs/maintainer-guide/issues#accepting-issues), should probably match the existing one for `null` (though, I'm a bit confused about `"never"`).

**Comment by SyMind** [comment]
@g-plane This is just an example that someone like use undefined.

**Comment by SyMind** [comment]
@mdjermanovic You get my meaning!

**Comment by mdjermanovic** [comment]
> Why are you confused about "never"? Can you give me more detail?

In my opinion, always requiring `foo == null` instead of `foo === null` can introduce new bugs or hide bugs that exist due to unconsidered code paths. Nevertheless, since that option already exists for `null`, options for `undefined` should probably mirror options for `null` and thus have `"never"`, if this enhancement gets accepted.

**Comment by SyMind** [comment]
I do not think is problem for `undefined`, it's for `null` with `never`.

**Comment by ljharb** [comment]
Also note that the `undefined` literal can be redefined, everywhere but the global scope. `== null` is the only safe usage of `==`.

**Comment by kaicataldo** [comment]
I don't think this change should be made in core for a few reasons:

1. Since this is purely stylistic, it falls under our [new policy](https://eslint.org/blog/2020/05/changes-to-rules-policies).

2. If you're never using `null`, why not just use `a === undefined` everywhere? It would be safer and more explicit. The way I read this guideline is that you should never use `null` to represent a null value for APIs you control. That doesn't mean you wouldn't have to use `a === null` in your code, since there's a good chance you'll be interacting with someone else's code (even if it's just the Node builtin or DOM APIs) that may return `null`.

**Comment by SyMind** [comment]
This is purely stylistic, but now eslint can't control with some config. Used `a == undefined` not only when `a = undefined` and also for `a = undefined` .

**Comment by nzakas** [comment]
Agree with @kaicataldo

**Comment by lemonbob** [comment]
I agree with @SyMind, `== undefined` should be a permitted option, and indeed a recommended option. I don't agree with @kaicataldo for the following reasons:

1. If you don't want `== undefined`, just leave the option off - we all get the EsLint that suits our code.
2. Douglas Crockford advises to never use null in JS, in his words, "it is utterly broken" (due to the typeof "object" bug). By using `== undefined` to include testing for null, in the very few circumstances that null is returned, we can simply forget about null. See — [Douglas Crockford - JS the Better Parts @ 30 minutes](https://www.youtube.com/watch?v=XFTOG895C7c)
3. `Undefined` cannot be removed from code because following TC39 committee decisions, it was selected as the preferred base value for JS, only `undefined` triggers ES6 default parameters and defaults for object deconstruction. Null was essentially semi-deprecated.
4. There is growing community support for the removal of null from code — see — [Intent to stop using `null` in my JS code](https://github.com/sindresorhus/meta/issues/7) and [Why I banned null from my JS code](https://medium.com/@hbarcelos/why-i-banned-null-from-my-js-code-and-why-you-should-too-13df90323cfa)
5. Programmers should never be redefining `undefined`, it is a long standing rule of very bad programming. Attempting to prevent a minuscule number of reckless programmers from doing something they should not be doing is a weak argument. Besides, if one did redefine `undefined` most code formatters, e.g. VS code, will indicate a change in color of `undefined`, thus you would have to be both reckless and unobservant. (If one wanted to, one could overwrite almost every prototype method in JS, but we don't stop overriding prototypes in EsLint.)
6. The "if you are never using null" argument above is incorrect. The only way to never use `null` is to allow `== undefined` because many external API calls return null, as does `string.match()` and several other edge cases. By not allowing `== undefined` it forces the continued use of null throughout code, which is confusing, confusion means lack of clarity, lack of clarity is the biggest cause of bugs. Thus there would be more bugs not fewer. In fact, this kind of thinking leads to the increased use of loose equality throughout as controversially espoused by Kyle Simpson.
7. Tony Hoare termed "null" the billion dollar mistake for a reason — [Null Billion Dollar Mistake](https://hackernoon.com/null-the-billion-dollar-mistake-8t5z32d6) 
8. In JS null will coerce to zero, (for no reason other than C permitted null pointers in mathematics equations). Thus null is unsafe for any form of numerical comparison using greater or less than. Undefined will never coerce in these situations and is void(0), a true base value. Thus again, removing null from code should be recommended and is only possible with `== undefined` as an option.

**Comment by kaicataldo** [comment]
It seems like there's a misunderstanding about what this rule is intended to enforce. `eqeqeq` is not recommending using `null` or `undefined` in general. The `allow-null` option exists so that folks can use the idiomatic shorthand `a == null` instead of having to write out `a === null || a === undefined`. The rule doesn't enforce setting any `null` values - it's just for this check.

I can see why folks want to try coalesce around a single nullish value, but this feels like a misapplication of this recommendation. Just because the code doesn't have the keyword `null` in it doesn't mean it isn't checking if the value is `null` or not. Since `a == null` is the convention, is safer, and one can always use `a === null || a === undefined` if they really don't want to use `a == null`, I don't think adding another option makes much sense. We should be encouraging the safer option.

The good news here is that ESLint is pluggable! If this conversation doesn't lead to implementing the requested feature in core, we encourage you to create your own custom rule ([eslint-rule-composer](https://www.npmjs.com/package/eslint-rule-composer) could be really useful here).

**Comment by ljharb** [comment]
@lemonbob there's a few inaccuracies in your comment; TC39 never "selected" a preferred base value for JS, `undefined` has *always* been the preferred base value since the language's inception. `null` is in no way deprecated by TC39 (which I am a member of). What programmers "should do" isn't the point; because they *can* redefine it, robust code needs to be concerned with it. As for coercion, if you're arguing what programmers should do, then one thing they shouldn't do is implicitly coerce anything :-)

`== undefined` should absolutely not be permitted by this rule, and imo a clean codebase would only ever use `== null` (or `!= null`), or `===`.

**Comment by lemonbob** [comment]
@ljharb your comments do not reflect the statements that Brendan Eich made re: undefined versus null. There was a discussion within TC39 of `null` acting as defaults for ES6 parameters in addition to `undefined`, but this was rejected for exactly the reasons I gave. Perhaps this was before your time on the committee, it was pre-2015? 

Brendan Eich wrote - "We have reason to believe typeof null === “object” is a bug that could bite real content, from our spidering of the web. It might be best to leave typeof utterly alone and deprecate [null]."

The purpose of EsLint rules is to "help teach users the correct way to use new language features".  

As, I am sure you are aware, the only way to achieve object optimisation and JIT compilation for object property access is for identity between objects. By that, I mean, identical in property names and identical order. To achieve this, it is common practice to define properties with base values to ensure objects are identical. Occasionally it is necessary, to utilise default parameters, to initialise these base values as empty. Because of `null` being utterly broken, this has led to the initialisation of some properties as `undefined`.

I agree that programmers should not implicitly coerce anything, but it is inevitable in any enterprise size Javascript codebase - unless you are arguing that code never has bugs! The selection of `== undefined` and never using `null` is simply one way, and a robust one as outlined by Douglas Cockcroft, to ensure programs are not affected by this very serious bug. 

As for the redefinition issue - EsLint has a `no console` rule - yet we can redefine console prototypes to be a mathematics sin method if we wanted, or any other essential method. It would be madness to do so, but you can. Should EsLint remove the `no console` rule just because a programmer could redefine it? Of course not! That is a straw man argument, as is the argument that the ability to redefined `== undefined` should preclude it from being a robust rule. 

@kaicataldo I appreciate that the `allow-null` option exists so that folks can use the idiomatic shorthand a `== null` instead of having to write out a === null || a === undefined. However, it is for this very reason that it promotes a coding pattern that is "utterly broken" (to use Douglas Cockcroft's words). To explain, it is very common practice to test for null and undefined in shorthand, furthermore this test is one of the most common tests in most code bases. To force users to write `=== null || === undefined` everywhere is not practical, thus by only supporting `== null` it implicitly promotes `null` as the preferred base value, especially for new programmers - this is the very thing contrary JS and contrary to EsLint policy.

**Comment by ljharb** [comment]
@lemonbob "it might be best to avoid X" is not "the official position of the committee is that X is deprecated", and it's very inaccurate to misrepresent as if it is. That a few vocal industry personalities have said something is "utterly broken" does not mean that the wider community and ecosystem agree. Use of null is in *no way* contrary to JS, or to "how the JS ecosystem uses these values" - also note that `null` is an exceedingly common value to use in the web platform which browsers *broadly encourage use of* by continuing to add APIs that accept and/or return `null`.

**Comment by SyMind** [comment]
@kaicataldo 
I never said i never using null, and If you're never using undefined, why not just use a === null everywhere, why not ban `"null": "always" | "never" | "ignore"` option.

@ljharb 
"it avalible to use undefined like `foo == undefined`" is not "i only use `undefined`",and it's very inaccurate to misrepresent as if it is. That a few vocal industry personalities have said something is "utterly broken" does not mean that the wider community and ecosystem agree. Use of undefined is in no way contrary to JS.

**Comment by ljharb** [comment]
If you do anything on the web, you deal with many APIs, some of which return `undefined` and some which return `null`, including in JS itself (`''.match(/./)`, eg). It's impossible to only ever use one of them - and in the world of modern web dev, most of your code is from dependencies, which you can't/shouldn't lint at all, and any of these might return whichever nullish value you hate the most. The purpose of this rule is to prevent all usages of `==`/`!=`, and the choice to make an exception for `== null` was, i assume, because that was both the common idiom, and also more robust than `== undefined`.

**Comment by lemonbob** [comment]
There are only two logical approaches here. 

1) EsLint should scrap the `allow null` rule and neither `== null` nor `== undefined` should be permitted. 
2) Both should be included, and if there was any real concern (which I would argue is nonsense) about developers redefining `undefined`, a separate linting rule warning against redefinition of undefined should be included. 

Indeed, given the serious warning from MDN, and the serious problems that would be caused by redefining `undefined`, that should be a rule in EsLint regardless! If a developed redefined `undefined` would that make any difference to `== undefined` being a less robust code? Of course not, the two are separate issues! If a developer redefined `undefined` they have more to worry about that a linting option, their code would be impossible to debug!

@ljharb When the creator of Javascript, Brendan Eich, writes that there is reason to believe typeof null is a bug, it is not a misrepresentation to state that there is wide support for concern over the use of null. Null was specifically excluded from ES6 defaults by the TC39 committee despite robust attempts to have it included - this is fact.

Regardless, I think we are digressing from the point.

As @SyMind  correctly stated, ECMA262 states that `null` is only to be used as the absence of an object - `undefined` is the absence of any value (including an object). Thus, `null` in ECMA262 script is a specific instance of `undefined`, a specific instance that is widely accepted to be buggy. That APIs return `null` is not a "recommendation" to use `null`, it is merely ECMA262 specification that it far too this late to change. The argument that it is a recommendation is akin to stating that gas stations selling fuel is a recommendation that we should use gas rather than electric cars - it's a straw man argument. Only humans can make recommendations, not browser APIs. The only internal JS method that returns null is string.match(), again, we can't change that, but JS programmers can follow advice from highly experienced programmers to never use `null`.

The facts are as follows:

1) Null is one of the most widely misused values in JS - just look at Stackoverflow - there are numerous non-ESMA262 compliant statements. Even the 2011 edition of `Javascript - The Definitive Guide`, advised programmers to set `Strings`, `Numbers` and `Booleans` to `null` instead of `undefined` if an empty value was required. This is corrected in the 2021 edition to be released.
2) Why was `allow null` included as a linting option? To make it easier to test for both `undefined` and `null` because as you correctly state, null and undefined are both returned - `null` only in the case of Objects. 
3) Does it make any sense for the preferred base value, which you accept is `undefined` to be excluded from the the linting option `== undefined`? Of course not.
4) [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/undefined) warns strongly against redefining `undefined`. 
5)Does allow null `== null` without `== undefined` cause more or less confusion and clarity - the most serious threat to robust code - of course it does, especially given point 1 above.
6) Given that there is a very strong warning from MDN not to redefine `undefined`, is it logical to cause more lack of clarity by making an illogical rule to exclude `== undefined` thus still allowing users to redefine `undefined`, or would it be more logical to add a linting rule to warn about the redefinition of `undefined` as well as a `== undefined` option.
7) It is impossible to exclude undefined from JS code because it is set as default, further it is required for ES6 features. However, as numerous, dozens of very experience developers (not very few - I gave three links alone) have very articulately stated - it is actually entirely possible to completely forget that null exists in JS because one can test for it by `== undefined`, further, null is only returned by a handful of methods, and many API calls.
8) It is well known, and stated by the one of the fathers of modern computer science, Tony Hoare, that null is the source for more bugs that almost anything else. In JS, this is made worse by the `typeof === 'object'` bug. Thus avoidance, if possible, should be an entirely recommended goal.

**Comment by ljharb** [comment]
@lemonbob i can assure you, again, that TC39 is a much larger body than one member, and
> Null was specifically excluded from ES6 defaults by the TC39 committee despite robust attempts to have it included - this is fact.

What robust attempts are you referring to? How do those attempts lead you to the conclusion that null is "deprecated"?

**Comment by lemonbob** [comment]
I didn't state that null _was_ deprecated, I stated that it is, _de facto_, by the recommendations and concerns of many of the founders of JS, including Eich himself, as "semi-deprecated". I will find the links if you want, but there were robust attempts to have null included in the ES6 spec to act akin to undefined for the new default parameters and object deconstruction defaults - those attempts were rejected by TC39 for many reasons, one of which was serious concern over the null bugs. 

@ljharb the above isn't really relevant to the discussion on `== undefined`, however, because it is enough to state that it is widely accepted that null is buggy. Also, as stated above, EsLints goal is to promote clarity and good code practice - only `== undefined` has the correct ECMA262 meaning to included both a test objects and non-objects when testing for emptiness, to include `== null` as the only exception is incorrect in terms of ECMA262 spec, it implies a test of absence of objects and will make code really unclear.

However, I take your point, redefinition of `undefined` isn't a good idea, for that reason EsLint should add that as a rule too, but that discussion is separate from `== undefined`, I think.

**Comment by ljharb** [comment]
I don't agree - `null` isn't buggy. `typeof null = 'object'` has a sensible explanation, retroactive tho it may be. As one of the editors of the ECMA262 specification, _I can assure you_ that `== null` is QUITE CORRECT in terms of that spec, both conceptually and in that the specification holds *no authority whatsoever* over a linter, since a linter doesn't run JS code.

> EsLint should add that as a rule too

You can't lint against something that third-party code might do. The only option is to prevent the damage that doing that would cause *everywhere* - which is why `== undefined` is a discouraged, arguably even deprecated pattern, and why `== null` is the only other sensible option.

**Comment by SyMind** [comment]
> Also note that the undefined literal can be redefined, everywhere but the global scope. == null is the only safe usage of ==.

In the ECMA262 specification, [[Writable]] of `undefined` is `false`. So i don't think is problem to use `undefined` in 2020.

> The value of undefined is undefined (see 6.1.1). This property has the attributes { [[Writable]]: false, [[Enumerable]]: false, [[Configurable]]: false }.

And most people use feature like ES Module and Class, they are always in strict mode. In strict mode will throw TypeError when `undefined = anything`.

**Comment by nzakas** [comment]
Thanks for the discussion, everyone. There isn't enough support on the team to implement this change, so we are closing this issue. As a reminder, you do have several options for getting this functionality including the use of the [`no-restricted-syntax`](https://eslint.org/docs/rules/no-restricted-syntax) rule or creating your own (either with or without [eslint-rule-composer](https://github.com/not-an-aardvark/eslint-rule-composer).

**Comment by lemonbob** [comment]
@ljharb as one of the editor of ECMA262, you will know that `null` is defined as follows:

4.4.15 null value
primitive value that represents the **intentional** absence of any **object** value

Please tell us, how does using `== null` to test for an _unintentional_ absence of a _non-object_ make any sense? The question is rhetorical, it doesn't make any logical or linguistic sense whatsoever. It couldn't be more wrong. Any editor would know this, and I used to be an assistant editor myself.

However, using `== undefined` to test for both intentional and unintentional absences of all types makes perfect sense because that is exactly what `undefined` is.

`== null` is a dangerous rule that conflates null with undefined, the two are not the same, and as such it does the very opposite of EsLint rules, it encourages bad code practice.

@SyMind and myself have gone to great lengths to provide refenece and supporting documentation of everything we have argued, all you keep relying on is _"I'm on the TC39 committee, I am editor of ECMA262."_ You have provided no shred of supporting evidence. If `== undefined` is deprecated please provide proof,

**Comment by ljharb** [comment]
@lemonbob i'm not going to argue further here, per https://github.com/eslint/eslint/issues/13537#issuecomment-677824969. I'm only using my TC39 status to object to your claim that TC39 has "deprecated" anything to do with null - please do not make that claim. `undefined` remains writable everywhere but in the global scope, and as such, `var undefined = true; x == undefined` will fail, and it's still a best practice to avoid use of the `undefined` literal.

**Comment by lemonbob** [comment]
As stated many times @ljharb, there is communication from Bredan Eich himself that TC39 rejected null in ES6. You were probably still in school when this happened, so I hope you will take the advice of those who actually wrote the language and were there when these decision were made.

Clearly null cannot be deprecated because too many programs now rely on this buggy bit of code, but that has been a discussed and expressed official desire. 

Incidentally, even the MDN reference makes use of the `undefined` literal in numerous examples, so please don't make claims without any evidence, contrary to actual evidence.
